### PR TITLE
Support HDFS rack locality

### LINKED ISSUE #349: Support HDFS rack locality
This is likely the last sub-item of HDFS locality umbrella issue #206.

When using HDFS, Spark driver looks up which rack a given datanode or executor belongs to.  So that it sends tasks to right executors that can read task data from datanodes on same racks. (This happens as a fallback when node locality fails). To support rack locality, Spark driver loads a configurable topology plugin into its JVM.

[TaskSchedulerImpl](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala) has the `getRackForHost` method, that is supposed to be overridden by a subclass to call the topology plugin. Then [TaskSetManager](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala) will call `getRackForHost` to populate `pendingTasksForRack` map. It gets called for each datanode host associated with input data blocks of pending tasks.

```
  private def addPendingTask(index: Int) {

...

      for (rack <- sched.getRackForHost(loc.host)) {
        pendingTasksForRack.getOrElseUpdate(rack, new ArrayBuffer) += index
      }
    }  
```

`getRackForHost` is also called with executor addresses when the driver is about to send tasks to executors.
```
  private def dequeueTask(execId: String, host: String, maxLocality: TaskLocality.Value)
    : Option[(Int, TaskLocality.Value, Boolean)] =

   ...

    if (TaskLocality.isAllowed(maxLocality, TaskLocality.RACK_LOCAL)) {
      for {
        rack <- sched.getRackForHost(host)
        index <- dequeueTaskFromList(execId, host, getPendingTasksForRack(rack))
      } {
        return Some((index, TaskLocality.RACK_LOCAL, false))
      }
    }
```

Yarn implements `getRackForHost` method in [YarnScheduler](https://github.com/apache-spark-on-k8s/spark/blob/branch-2.1-kubernetes/yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnScheduler.scala):

```
  override def getRackForHost(hostPort: String): Option[String] = {
    val host = Utils.parseHostPort(hostPort)._1
    Option(RackResolver.resolve(sc.hadoopConfiguration, host).getNetworkLocation)
  }
```

We can add similar code in [KubernetesTaskSchedulerImpl](https://github.com/apache-spark-on-k8s/spark/blob/branch-2.1-kubernetes/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala). The datanode handling will be exactly like the above. For executors, we would map pod IP addresses to cluster node name/IP and then call the topology plugin using the cluster node address. I'll send a PR soon.

### PR DESCRIPTION
**Pull request by kimoonkim**
Closes #349 and #206.

@ash211 @foxish 

Supports HDFS rack locality by implementing `getRackForHost` in [KubernetesTaskSchedulerImpl](https://github.com/apache-spark-on-k8s/spark/blob/branch-2.1-kubernetes/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala)

Added unit tests.

Also did manual testing using a dummy topology script that always returns a dummy rack name, "/rack0".

The driver log shows a small number of`RACK_LOCAL` tasks, which used to be `ANY` tasks. (The majority of tasks are still `NODE_LOCAL` tasks)

>2017-06-16 16:56:38 INFO  KubernetesTaskSetManager:54 - Starting task 34.0 in stage 0.0 (TID 57, 10.44.0.5, executor 9, partition 34, RACK_LOCAL, 6718 bytes)
2017-06-16 16:56:39 INFO  KubernetesTaskSetManager:54 - Starting task 40.0 in stage 0.0 (TID 59, 10.44.0.4, executor 7, partition 40, RACK_LOCAL, 6718 bytes)
2017-06-16 16:56:39 INFO  KubernetesTaskSetManager:54 - Starting task 50.0 in stage 0.0 (TID 63, 10.46.0.4, executor 5, partition 50, RACK_LOCAL, 6719 bytes)

The job was `HdfsTest`.

>/usr/local/spark-on-k8s/bin/spark-submit --class org.apache.spark.examples.HdfsTest --conf spark.app.name=spark-hdfstest --conf spark.dynamicAllocation.enabled=false --conf spark.shuffle.service.enabled=false --conf spark.executor.instances=10 --conf spark.kubernetes.shuffle.labels="app=spark-shuffle-service,spark-version=2.1.0" local:///opt/spark/examples/jars/spark-examples_2.11-2.1.0-k8s-0.2.0-SNAPSHOT.jar 10GB-txt

`spark-defaults.conf` specified the dummy topology script:
```
spark.hadoop.fs.defaultFS hdfs://hdfs-namenode-0.hdfs-namenode.default.svc.cluster.local:8020
spark.hadoop.net.topology.script.file.name /tmp/print_rack.sh
```

The dummy script `print_rack.sh`
```
#!/bin/bash

echo /rack-0
```

I added the script to the driver docker image manually.

### CODE REVIEW

**Comment by kimoonkim** [comment]
FYI, the unit tests failure seem genuine. I'm looking at it.

**Comment by foxish** [comment]
Thanks @kimoonkim! This looks awesome. Looking into this in detail shortly.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala [hunk 1]
```diff
@@ -16,12 +16,97 @@
...
+  val scriptPlugin = classOf[ScriptBasedMapping].getCanonicalName
+  val tablePlugin = classOf[TableMapping].getCanonicalName
+
+  var isConfigured = false
+
+  def init(hadoopConfiguration: Configuration) : Unit = {
+    isConfigured = checkConfigured(hadoopConfiguration)
+    // RackResolver logs an INFO message whenever it resolves a rack, which is way too often.
+    if (Logger.getLogger(classOf[RackResolver]).getLevel == null) {
+      Logger.getLogger(classOf[RackResolver]).setLevel(Level.WARN)
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala]: haha yes I've seen this too -- I think Hadoop 2.8.0 lowered this log level.  What version of are you testing against?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala [hunk 2]
```diff
@@ -16,12 +16,97 @@
...
+import org.apache.log4j.{Level, Logger}
+
 import org.apache.spark.scheduler.{TaskSchedulerImpl, TaskSet, TaskSetManager}
+import org.apache.spark.util.Utils
+import org.apache.spark.SparkContext
 
-private[spark] class KubernetesTaskSchedulerImpl(sc: SparkContext) extends TaskSchedulerImpl(sc) {
+private[spark] class KubernetesTaskSchedulerImpl(
+    sc: SparkContext,
+    rackResolverUtil: RackResolverUtil = new RackResolverUtil,
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala]: can you put sc as a parameter into a RackResolverUtil constructor?  I'm hoping to be able to get rid of the `rackResolverUtil.init` method since that class has a two-step initialization step with it (instantiate class, call `.init`)

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala [hunk 3]
```diff
@@ -16,12 +16,97 @@
...
+      // See InetAddressUtil for details.
+      None
+    } else {
+      getRackForDatanodeOrExecutor(hostPort)
+    }
+  }
+
+  private def getRackForDatanodeOrExecutor(hostPort: String): Option[String] = {
+    val host = Utils.parseHostPort(hostPort)._1
+    val backend = this.backend.asInstanceOf[KubernetesClusterSchedulerBackend]
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala]: do this cast once in the constructor (outside this method) ?  if this is cheap in the JVM JIT then maybe no need to avoid frequent casts

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala [hunk 4]
```diff
@@ -16,12 +16,97 @@
...
+    val executorPod = backend.getExecutorPodByIP(host)
+    if (executorPod.isEmpty) {
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala]: this if/else could be turned into a `match` which might be cleaner in scala

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala [hunk 5]
```diff
@@ -16,12 +16,97 @@
...
+    inetAddressUtil: InetAddressUtil = new InetAddressUtil) extends TaskSchedulerImpl(sc) {
+
+  rackResolverUtil.init(sc.hadoopConfiguration)
 
   override def createTaskSetManager(taskSet: TaskSet, maxTaskFailures: Int): TaskSetManager = {
     new KubernetesTaskSetManager(this, taskSet, maxTaskFailures)
   }
+
+  override def getRackForHost(hostPort: String): Option[String] = {
+    if (!rackResolverUtil.isConfigured) {
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala:42]: nice work on keeping this speedy for non-HDFS users

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSetManagerSuite.scala [hunk 6]
```diff
@@ -21,12 +21,13 @@ import scala.collection.mutable.ArrayBuffer
...
 
-import org.apache.spark.{SparkContext, SparkFunSuite}
+import org.apache.spark.{SparkConf, SparkContext, SparkFunSuite}
 import org.apache.spark.scheduler.{FakeTask, FakeTaskScheduler, HostTaskLocation, TaskLocation}
 
 class KubernetesTaskSetManagerSuite extends SparkFunSuite {
 
-  val sc = new SparkContext("local", "test")
+  val sc = new SparkContext(master = "local", appName = "test",
+    new SparkConf().set("spark.driver.allowMultipleContexts", "true"))
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSetManagerSuite.scala]: why is this change needed?  are we parallelizing tests and running multiple drivers in the same JVM at once?

**Review by ash211 — COMMENTED** [review]
Nice work @kimoonkim !

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala (hunk 2 above)
Comment (kimoonkim) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala]: Fixed. Thanks for the suggestion. I wanted to do this earlier but didn't try hard enough :-)

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala (hunk 3 above)
Comment (kimoonkim) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala]: Fixed. FYI it appears this has to be done in the separate init method.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala (hunk 4 above)
Comment (kimoonkim) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala]: Fixed.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala (hunk 1 above)
Comment (kimoonkim) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala]: I was using 2.7. Good to know 2.8 fixed this ([YARN-3350](https://issues.apache.org/jira/browse/YARN-3350))

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSetManagerSuite.scala (hunk 6 above)
Comment (kimoonkim) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSetManagerSuite.scala]: In this project, we have `KubernetesTaskSetManagerSuite` and `KubernetesTaskSchedulerImplSuite`. The two suites run inside the single JVM. We are still running one suite at a time sequentially. But we are creating two `SparkContext` instances. Without this option, the second suite fails because only one `SparkContext` instance is allowed. I was hoping to prevent that. 

But I think there is a better option, which is just to call `SparkContext.clearActiveContext()`. Switched to that.

**Review by foxish — APPROVED** [review]

**Comment by foxish** [comment]
LGTM! Thanks @kimoonkim, that looks good, and appears to handle both hdfs nodes and executor pods well. Question - for the executor pods, how do you see the script to resolve rack-info working? Could it have access to more than just the Pod IP to find what rack it belongs to?

**Comment by kimoonkim** [comment]
@foxish Good question. My understanding is that executors do not call the topology plugin. Only the driver will consult with the topology plugin to decide which executor, hopefully a rack-local one, should receive a new task.

When an executor reads a HDFS block for the new task, it will then simply use Hadoop library code that sends a RPC request to the namenode. The namenode will consider the list of datanodes that have copies of the block. And the namenode asks the topology plugin which datanodes are better. When the namenode returns the list of datanodes, it sort them in the locality order. (node local to rack local to remote)

FYI, the HDFS on kubernetes we have does not support configuration of the topology plugin yet in the namenode helm chart. But I intend to do it soon.

The pod IP question on the namenode side is not as important as we originally thought. Because most k8s network plugins do NAT and the namenode sees the k8s cluster node IPs. (The pod IP issue on the namenode side only manifests in the kubenet on GKE) For details, please see this [kubernetes-HDFS/topology/README.md](https://github.com/apache-spark-on-k8s/kubernetes-HDFS/blob/master/topology/README.md)

**Comment by kimoonkim** [comment]
@ash211 @foxish @mccheah Thanks for reviews so far. I wonder if we have any more questions or comments on this. Maybe ready to be merged soon?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala [hunk 7]
```diff
@@ -16,12 +16,99 @@
...
+      None
+    } else {
+      getRackForDatanodeOrExecutor(hostPort)
+    }
+  }
+
+  private def getRackForDatanodeOrExecutor(hostPort: String): Option[String] = {
+    val host = Utils.parseHostPort(hostPort)._1
+    val executorPod = kubernetesSchedulerBackend.getExecutorPodByIP(host)
+    executorPod.isEmpty match {
```
Comment (mccheah) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala]: Use `Option.map.getOrElse` instead of matching on `isEmpty`.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+import org.apache.hadoop.fs.CommonConfigurationKeysPublic
+import org.mockito.Matchers._
+import org.mockito.Mockito._
+
+import org.apache.spark.{SparkContext, SparkFunSuite}
+import org.apache.spark.scheduler.FakeTask
+
+class KubernetesTaskSchedulerImplSuite extends SparkFunSuite {
+
+  SparkContext.clearActiveContext()
```
Comment (mccheah) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala]: Should this be done in a `before` block?

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+  val sc = new SparkContext("local", "test")
```
Comment (mccheah) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala]: Can we just use a mock SparkContext here?

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    val sched = new KubernetesTaskSchedulerImpl(sc)
+    sched.kubernetesSchedulerBackend = backend
+    val taskSet = FakeTask.createTaskSet(0)
+
+    val manager = sched.createTaskSetManager(taskSet, maxTaskFailures = 3)
+    assert(manager.isInstanceOf[KubernetesTaskSetManager])
+  }
+
+  test("Gets racks for datanodes") {
+    val rackResolverUtil = mock(classOf[RackResolverUtil])
```
Comment (mccheah) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala:47]: It seems unusual to me to be mocking the `RackResolverUtil`. It seems to be part of the core `KubernetesTaskSchedulerImpl` because the `RackResolverUtil` is a nested private class.

If we indeed want to be testing these separately then the architecture should reflect as such:

- Extract `RackResolverUtil` to another class,
- Put a `trait` on top of the `RackResolverUtil`,
- Inject an instance into the `KubernetesTaskSchedulerImpl` when we create it,
- Write a separate test for `RackResolverUtil`.

If we don't want to test these separately then we should create a real `RackResolveUtil` and test the top level methods accordingly.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    when(rackResolverUtil.isConfigured).thenReturn(true)
+    when(rackResolverUtil.resolveRack(sc.hadoopConfiguration, "kube-node1"))
+      .thenReturn(Option("/rack1"))
+    when(rackResolverUtil.resolveRack(sc.hadoopConfiguration, "kube-node2.mydomain.com"))
+      .thenReturn(Option("/rack2"))
+    when(rackResolverUtil.resolveRack(sc.hadoopConfiguration, "kube-node2"))
+      .thenReturn(None)
+    when(rackResolverUtil.resolveRack(sc.hadoopConfiguration, "192.168.1.5"))
+      .thenReturn(None)
+    val inetAddressUtil = mock(classOf[InetAddressUtil])
```
Comment (mccheah) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala:74]: Any reason this can't be a real `InetAddressUtil`?

Comment (mccheah) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala:74]: If it can't be then move `InetAddressUtil` to its own file and place a `trait` over it. The real `InetAddressUtil` could probably be an `object` that extends the trait in this case.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSetManagerSuite.scala [hunk 12]
```diff
@@ -26,6 +26,7 @@ import org.apache.spark.scheduler.{FakeTask, FakeTaskScheduler, HostTaskLocation
 
 class KubernetesTaskSetManagerSuite extends SparkFunSuite {
 
+  SparkContext.clearActiveContext()
```
Comment (mccheah) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSetManagerSuite.scala]: Again - should this be being done in a `before` block? Also, can we use a mock `SparkContext`?

**Review by mccheah — CHANGES_REQUESTED** [review]
Sorry this took awhile to review. I have some suggestions.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala (hunk 7 above)
Comment (kimoonkim) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImpl.scala]: Fixed. Thanks for the suggestion. Found and fixed a minor bug thanks to that!

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala (hunk 8 above)
Comment (kimoonkim) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala]: It could be put in a `before` block. But then I need to put `val sc = new SparkContext(...)` also inside the block because `SparkContext.clearActiveContext()` should be called before a new `SparkContext` is created. Please let me know what you think.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala (hunk 9 above)
Comment (kimoonkim) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala]: Yeah, I thought about the possibility and tried a bit. But then I realized `SparkContext` is not friendly to mock. It just has too many methods returning too many objects. I'll have to mock many of those that the test subject class happens to interact with. And if any of interaction changes, the test will break.

I think the test maintenance cost becomes too much. I found other Spark core unit tests avoid using mock probably for the same reason.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala (hunk 10 above)
Comment (kimoonkim) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala:47]: Good point. It started as a pure wrapper of RackResolver, then I ended up adding a few business logics. I like the suggestion, but I'll have to think about this a little bit.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala (hunk 11 above)
Comment (kimoonkim) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala:74]: I wanted to mock it so unit tests don't call real DNS and potentially get influenced by the responses. The trail approach sounds good, I'll probably try in the next patch.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSetManagerSuite.scala (hunk 12 above)
Comment (kimoonkim) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSetManagerSuite.scala]: Please see the answer above.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala (hunk 10 above)
Comment (kimoonkim) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala:47]: Fixed. New code looks better. Thanks.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala (hunk 11 above)
Comment (kimoonkim) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSchedulerImplSuite.scala:74]: Fixed.

**Comment by kimoonkim** [comment]
@mccheah Thanks for the review. Addressed comments. PTAL.

**Comment by kimoonkim** [comment]
I was trying to retarget this to branch-2.2-kubernetes using github UI. Then I realized that there will be too many diffs in this PR.

Maybe I should retarget to branch-2.2-kubernetes and merge changes from branch-2.2-kubernetes? @foxish Any better suggestion?

**Comment by kimoonkim** [comment]
Ok. Retargeted to branch-2.2-kubernetes after rebasing my branch.

**Comment by kimoonkim** [comment]
Merged with `branch-2.2-kubernetes`. While at it, also incorporated the flag from #412 to avoid expensive DNS lookup.

**Comment by kimoonkim** [comment]
@mccheah Can you please take a look at this PR? Perhaps, this is ready to merge after another look.

**Review by mccheah — APPROVED** [review]

**Comment by kimoonkim** [comment]
@foxish @ash211 @mccheah Thanks for the reviews. Maybe this can be merged now?
